modules/rostests/apitests/shell32/ExtractIconEx.cpp

```diff
@@ -0,0 +1,40 @@
+/*
+ * PROJECT:     ReactOS API Tests
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Tests for ExtractIconEx routine
+ * COPYRIGHT:   Copyright 2019 Bișoc George (fraizeraust99 at gmail dot com)
+ */
+
+#include "shelltest.h"
+
+typedef struct
+{
+    PCWSTR pszFilePath;
+    UINT nIcons;
+    BOOL bFileExists;
+} EXTRACTICONTESTS;
+
+EXTRACTICONTESTS IconTests[] =
+{
+    /* Executable file with icon */
+    {L"%SystemRoot%\\System32\\cmd.exe", 1, TRUE},
+
+    /* Executable file without icon */
+    {L"%SystemRoot%\\System32\\autochk.exe", 0, TRUE},
+
+    /* Non-existing files */
+    {L"%SystemRoot%\\non-existent-file.sdf", 0, FALSE},
+};
+
+START_TEST(ExtractIconEx)
+{
+    UINT i, nReturnedIcons;
+
+    for (i = 0; i < _countof(IconTests); ++i)
+    {
+        nReturnedIcons = ExtractIconExW(IconTests[i].pszFilePath, 0, NULL, NULL, IconTests[i].nIcons);
+        ok(nReturnedIcons == IconTests[i].nIcons, "ExtractIconExW(%u): Expected %u icons, got %u\n", i, IconTests[i].nIcons, nReturnedIcons);
+        ok((IconTests[i].bFileExists && nReturnedIcons == IconTests[i].nIcons) || (!IconTests[i].bFileExists && nReturnedIcons == IconTests[i].nIcons),
```

Comment:
I don't know what your intention with this check is, but the logic can be simplified:
```
(a && b) || (!a && b)
===
(a || !a) && b
===
b
```

So all you're checking here is `nReturnedIcons == IconTests[i].nIcons`, which is the same as the line above.

Reply:
The core logic behind this is as follows:

- If a file couldn't be found because it doesn't exist, logically you return 0 as in no icons could be extracted due to the file not being found. However a file may not necessarily have an icon so you just return 0 (in case the file in question exists).

Now that I am talking about this, the extra line somewhat achieves this purpose as the same line above. I guess I shall remove that line.